packages/mui-core/src/InputUnstyled/InputUnstyled.tsx

```diff
@@ -0,0 +1,281 @@
+    'aria-label': ariaLabel,
+    'aria-labelledby': ariaLabelledby,
+    autoComplete,
+    autoFocus,
+    className,
+    component,
+    components = {},
+    componentsProps = {},
+    defaultValue,
+    disabled,
+    error,
+    id,
+    maxRows,
+    minRows,
+    multiline = false,
+    name,
+    onKeyDown,
+    onKeyUp,
+    placeholder,
+    readOnly,
+    rows,
+    type = 'text',
+    value,
+    ...other
+  } = props;
+
+  const {
+    getRootProps,
+    getInputProps,
+    focused,
+    formControlContext,
+    error: errorState,
+    disabled: disabledState,
+  } = useInput(props);
+
+  const ownerState = {
+    ...props,
+    disabled: disabledState,
+    error: errorState,
+    focused,
+    formControlContext,
+    multiline,
+    type,
+  };
+
+  const rootStateClasses = {
+    [classes.disabled]: disabledState,
+    [classes.error]: errorState,
+    [classes.focused]: focused,
+    [classes.formControl]: formControlContext,
+    [classes.multiline]: multiline,
+  };
+
+  const inputStateClasses = {
+    [classes.disabled]: disabledState,
+    [classes.multiline]: multiline,
+  };
+
+  const propsToForward = {
+    'aria-describedby': ariaDescribedby,
+    'aria-label': ariaLabel,
+    'aria-labelledby': ariaLabelledby,
+    autoComplete,
+    autoFocus,
+    id,
+    onKeyDown,
+    onKeyUp,
+    name,
+    placeholder,
+    readOnly,
+    type,
+  };
+
+  const Root = component ?? components.Root ?? 'div';
+  const rootProps = appendOwnerState(
+    Root,
+    {
+      ...other,
+      ...getRootProps(),
+      ...componentsProps.root,
+      className: clsx(classes.root, rootStateClasses, className, componentsProps.root?.className),
+    },
+    ownerState,
+  );
+
+  let Input = components.Input ?? 'input';
+  let inputProps = appendOwnerState(
+    Input,
+    {
+      ...propsToForward,
+      ...getInputProps(),
+      ...componentsProps.input,
+      className: clsx(classes.input, inputStateClasses, componentsProps.input?.className),
+    },
+    ownerState,
+  );
+
+  if (multiline) {
+    const hasHostTexarea = isHostComponent(components.Textarea ?? 'textarea');
+
+    if (rows) {
+      if (process.env.NODE_ENV !== 'production') {
+        if (minRows || maxRows) {
+          console.warn(
+            'Material-UI: You can not use the `minRows` or `maxRows` props when the input `rows` prop is set.',
+          );
+        }
+      }
+      inputProps = {
+        type: undefined,
+        minRows: hasHostTexarea ? undefined : rows,
+        maxRows: hasHostTexarea ? undefined : rows,
+        ...inputProps,
+      };
+    } else {
+      inputProps = {
+        type: undefined,
+        maxRows: hasHostTexarea ? undefined : maxRows,
+        minRows: hasHostTexarea ? undefined : minRows,
+        ...inputProps,
+      };
+    }
+
+    Input = components.Textarea ?? 'textarea';
+  }
+
+  return (
+    <Root {...rootProps} ref={ref}>
+      <Input {...inputProps} />
+    </Root>
```

Comment:
Another question, do we really need `Root`? can it be only `<Input />`

Reply:
Root is useful when you have adornments. I suppose they should be a part of the unstyled control as well, after all. I'm going to add them.